InvenTree/plugin/apps.py

```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+import importlib
+import pathlib
+import logging
+from typing import OrderedDict
+
+from django.apps import AppConfig, apps
+from django.conf import settings
+from django.db.utils import OperationalError, ProgrammingError
+
+try:
+    from importlib import metadata
+except:
+    import importlib_metadata as metadata
+
+from plugin import plugins as inventree_plugins
+from plugin.integration import IntegrationPluginBase
+
+logger = logging.getLogger('inventree')
+
+
+class PluginAppConfig(AppConfig):
+    name = 'plugin'
+
+    def ready(self):
+        self.collect_plugins()
+
+        try:
+            # we are using the db from here - so for migrations etc we need to try this block
+            self.init_plugins()
+            self.activate_integration()
+        except (OperationalError, ProgrammingError):
+            # Exception if the database has not been migrated yet
+            pass
+
+    def collect_plugins(self):
+        """collect integration plugins from all possible ways of loading"""
+        # Collect plugins from paths
+        for plugin in settings.PLUGIN_DIRS:
+            modules = inventree_plugins.get_plugins(importlib.import_module(plugin), IntegrationPluginBase, True)
+            if modules:
+                [settings.PLUGINS.append(item) for item in modules]
+
+        # Collect plugins from setup entry points
+        for entry in metadata.entry_points().get('inventree_plugins', []):
+            plugin = entry.load()
+            plugin.is_package = True
+            settings.PLUGINS.append(plugin)
+
+        # Log found plugins
+        logger.info(f'Found {len(settings.PLUGINS)} plugins!')
+        logger.info(", ".join([a.__module__ for a in settings.PLUGINS]))
+
+    def init_plugins(self):
+        """initialise all found plugins"""
+        from plugin.models import PluginConfig
+
+        logger.info('Starting plugin initialisation')
+        # Initialize integration plugins
+        for plugin in inventree_plugins.load_integration_plugins():
+            # check if package
+            was_packaged = getattr(plugin, 'is_package', False)
+
+            # check if activated
+            # these checks only use attributes - never use plugin supplied functions -> that would lead to arbitrary code execution!!
+            plug_name = plugin.PLUGIN_NAME
+            plug_key = plugin.PLUGIN_SLUG if getattr(plugin, 'PLUGIN_SLUG', None) else plug_name
+            plugin_db_setting, _ = PluginConfig.objects.get_or_create(key=plug_key, name=plug_name)
+
+            # always activate if testing
+            if settings.PLUGIN_TESTING or plugin_db_setting.active:
```

Comment:
@SchrodingersGat any idea why this will not work? I am a bit stuck here

Reply:
@matmair what error message are you seeing? What are you trying to do?

Reply:
The tests fail because the plugins do not load. At this point the plugin should be loaded but it seems like the tests never reach the point.

Reply:
Ah right, I didn't see that the tests were failing at that point. Did you get it fixed?